sdk/iot/docs/mqtt_state_machine.md

```diff
@@ -0,0 +1,223 @@
+    az_iot_topic topic;
+    az_span payload;
+} az_iot_mqtt_pub;
+```
+
+### Client Initialization
+```C
+// Provisioning
+
+typedef struct az_provisioning_client {
+    struct {
+        az_span registration_id;
+    } _internal;
+} az_provisioning_client;
+
+void az_provisioning_client_init(az_provisioning_client* client, az_span registration_id);
+
+// IoT Hub
+
+typedef struct az_iot_hub_client {
+    struct {
+        az_span identity;   // deviceId[/moduleId]
+    } _internal;
+} az_provisioning_client;
+
+void az_iot_hub_client_init(az_iot_hub_client* client, az_span device_id);
+void az_iot_hub_client_init(az_iot_hub_client* client, az_span device_id, az_span module_id);
+```
+
+### Connect Credentials
+
+Two authentication schemes are currently supported: _X509 Client Certificate Authentication_ and _Shared Access Signature_ authentication. The API must support both cases as well without making assumptions of direct access to secrets (Hardware Security Modules are supported).
+
+```C
+// Provisioning
+az_result az_iot_provisioning_client_init(az_provisioning_client* client, az_span id_scope, az_iot_mqtt_connect* mqtt_connect);
+az_result az_iot_provisioning_get_connect_information(az_iot_hub_client* client, az_span id_scope, az_iot_mqtt_connect* mqtt_connect);
+
+// IoT Hub
+az_result az_iot_hub_client_init(az_iot_hub_client* client, az_iot_mqtt_connect* mqtt_connect);
+az_result az_iot_hub_get_connect_information(az_iot_hub_client* client, az_span hub_hostname, az_iot_mqtt_connect* mqtt_connect);
+```
+
+If X509 Client authenticaiton is not used, `mqtt_connect.password` must contain a valid SAS Token. The password is obtained by using 2 function calls: one to initialize a SAS token and another to sign and update it.
+
+```C
+// Provisioning
+typedef struct az_iot_provisioning_sas {
+    struct {
+        az_span audience; // URLENC("<id_scope>/registrations/<registration_id>"). This is HMAC256-ed together with expiration_time.
+        az_span key_name; // Can be empty.
+    } _internal;
+} az_iot_provisioning_sas;
+
+az_result az_iot_provisioning_sas_init(az_iot_provisioning_sas* sas, az_provisioning_client* client, az_span id_scope, az_span key_name);
+az_result az_iot_provisioning_sas_update(az_iot_provisioning_sas* sas, time_t expiration_time, az_span base64_hmac256_signature, az_iot_mqtt_connect* mqtt_connect);
+
+// IoT Hub
+typedef struct az_iot_hub_sas {
+    struct {
+        az_span audience;
+        az_span key_name;
+    } _internal;
+} az_iot_hub_sas;
+
+az_result az_iot_hub_sas_init(az_iot_hub_sas* sas, az_iot_hub_client* client, az_span hub_name, az_span key_name);
+az_result az_iot_hub_sas_update(az_iot_hub_sas* sas, time_t expiration_time, az_span base64_hmac256_signature, az_iot_mqtt_connect* mqtt_connect);
+```
+
+
+### Subscribe topic information
+Note: All services (Device Provisioning, IoT Hub) require either no subscription (Telemetry) or _a single_ subscription to a topic _filter_.

```

Comment:
I do not believe this is true. Twin has subscriptions to the following:
```
$iothub/twin/res/#
$iothub/twin/PATCH/properties/desired/#
```

Reply:
You're right: I forgot to fix the doc after our discussion. There should be 2 separate API calls to get the PATCH vs the GET response for Twin. These are missing.